### PR TITLE
refactor(first): make first smaller

### PR DESCRIPTION
**Pull request by benlesh**
Reduces the size of first my implementing it in terms of other operators.

### CODE REVIEW

File: src/operators/first.ts [hunk 1]
```diff
@@ -67,100 +75,34 @@ export function first<T>(predicate: (value: T, index: number, source: Observable
...
-  private _emitFinal(value: any) {
-    const destination = this.destination;
-    if (!this._emitted) {
-      this._emitted = true;
-      destination.next(value);
-      destination.complete();
-      this.hasCompleted = true;
-    }
-  }
+  const after = defaultValue ? of(defaultValue) : _throw(new EmptyError());
```
Comment (cartant) [inline src/operators/first.ts]: Testing whether or not `defaultValue` is falsy won't allow for default values of `0` or `false`, etc.

Comment (benlesh) [reply src/operators/first.ts]: Right. Don't code tired. Thanks

**Comment by martinsik** [comment]
I tested this PR and it actually degrades performance compared to the previous implementation but I understand that this operator isn't used as much so maybe it doesn't matter (it's cca 10x slower).

Now:
```
                                         |                     RxJS 4.1.0 |                     RxJS 5.5.3 |          factor |      % improved
---------------------------------------------------------------------------------------------------------------------------------------------------
                       first - immediate |                34,519 (±4.93%) |               253,469 (±4.53%) |           7.34x |          634.3%
```

Before:

```
                                         |                     RxJS 4.1.0 |                     RxJS 5.5.3 |          factor |      % improved
---------------------------------------------------------------------------------------------------------------------------------------------------
                       first - immediate |                27,327 (±5.23%) |             2,053,011 (±3.78%) |          75.13x |        7,412.9%
```

**Comment by benlesh** [comment]
@martinsik yeah.. I'm dubious about the change here myself. I'll look at a few other ways to reduce the size. This is more of an experiment.

**Pull request closed without merging**